File: sdk/java/src/main/java/com/obtbrowser/ProxyApi.java
package com.obtbrowser;

import com.fasterxml.jackson.databind.JsonNode;

import java.util.HashMap;
import java.util.Map;

/** Proxy pool management — CRUD, assign, and unassign proxies. */
public class ProxyApi {

    private final HttpHelper http;

    ProxyApi(HttpHelper http) {
        this.http = http;
    }

    /** List all proxies. */
    public JsonNode list() throws Exception {
        return http.get("/api/proxies");
    }

    /** List proxies that are not assigned to any profile. */
    public JsonNode listUnassigned() throws Exception {
        return http.get("/api/proxies/unassigned");
    }

    /**
     * Create a proxy.
     *
     * @param type one of: http, https, socks4, socks5
     */
    public JsonNode create(String type, String host, int port,
            String name, String username, String password) throws Exception {
        Map<String, Object> body = new HashMap<>();
        body.put("type", type);
        body.put("host", host);
        body.put("port", port);
        if (name != null)
            body.put("name", name);
        if (username != null)
            body.put("username", username);
        if (password != null)
            body.put("password", password);
        return http.post("/api/proxies", body);
    }

    /** Update proxy fields. Pass null for fields you do not want to change. */
    public JsonNode update(String proxyId, String name, String type, String host,
            Integer port, String username, String password) throws Exception {
        Map<String, Object> body = new HashMap<>();
        if (name != null)
            body.put("name", name);
        if (type != null)
            body.put("type", type);
        if (host != null)
            body.put("host", host);
        if (port != null)
            body.put("port", port);
        if (username != null)
            body.put("username", username);
        if (password != null)
            body.put("password", password);
        return http.put("/api/proxies/" + proxyId, body);
    }

    /** Delete a proxy. */
    public void delete(String proxyId) throws Exception {
        http.delete("/api/proxies/" + proxyId);
    }

    /** Assign a proxy to a profile. */
    public JsonNode assign(String proxyId, String profileId) throws Exception {
        return http.post("/api/proxies/" + proxyId + "/assign", Map.of("profileId", profileId));
    }

    /** Unassign any proxy currently attached to a profile. */
    public JsonNode unassign(String profileId) throws Exception {
        return http.post("/api/proxies/unassign", Map.of("profileId", profileId));
    }

    /** Clear stored proxy username and password. */
    public JsonNode clearCredentials(String proxyId) throws Exception {
        Map<String, Object> body = new HashMap<>();
        body.put("username", null);
        body.put("password", null);
        return http.put("/api/proxies/" + proxyId, body);
    }
}

File: sdk/python/obt_browser/proxy_api.py
"""Proxy pool operations — CRUD, assign, and unassign."""

from typing import Any, Optional

from ._http import BaseApi


class ProxyApi(BaseApi):
    def list(self) -> list:
        """Return all proxies."""
        return self._get("/api/proxies")

    def list_unassigned(self) -> list:
        """Return proxies that are not assigned to any profile."""
        return self._get("/api/proxies/unassigned")

    def create(
        self,
        proxy_type: str,
        host: str,
        port: int,
        name: Optional[str] = None,
        username: Optional[str] = None,
        password: Optional[str] = None,
    ) -> dict:
        """Create a proxy. proxy_type must be http/https/socks4/socks5."""
        body: dict[str, Any] = {
            "type": proxy_type,
            "host": host,
            "port": port,
        }
        if name is not None:
            body["name"] = name
        if username is not None:
            body["username"] = username
        if password is not None:
            body["password"] = password
        return self._post("/api/proxies", body)

    def update(
        self,
        proxy_id: str,
        name: Optional[str] = None,
        proxy_type: Optional[str] = None,
        host: Optional[str] = None,
        port: Optional[int] = None,
        username: Optional[str] = None,
        password: Optional[str] = None,
    ) -> dict:
        """Update one or more proxy fields."""
        body: dict[str, Any] = {}
        if name is not None:
            body["name"] = name
        if proxy_type is not None:
            body["type"] = proxy_type
        if host is not None:
            body["host"] = host
        if port is not None:
            body["port"] = port
        if username is not None:
            body["username"] = username
        if password is not None:
            body["password"] = password
        return self._put(f"/api/proxies/{proxy_id}", body)

    def delete(self, proxy_id: str) -> None:
        """Delete a proxy."""
        self._delete(f"/api/proxies/{proxy_id}")

    def assign(self, proxy_id: str, profile_id: str) -> dict:
        """Assign a proxy to a profile."""
        return self._post(f"/api/proxies/{proxy_id}/assign", {"profileId": profile_id})

    def unassign(self, profile_id: str) -> dict:
        """Unassign the current proxy from a profile."""
        return self._post("/api/proxies/unassign", {"profileId": profile_id})

    def clear_credentials(self, proxy_id: str) -> dict:
        """Clear stored proxy username and password."""
        return self._put(
            f"/api/proxies/{proxy_id}",
            {
                "username": None,
                "password": None,
            },
        )